Use Feature::Compat::Try everywhere

okurz opened the pull request:
* Use Feature::Compat::Try wherever possible
* t: Consistently use Test::Exception method instead of manual eval checks
* t: Fix check for simplified Test::Exception
* Use Feature::Compat::Try in place of Try::Tiny
* Remove unused 'try' import
* Add dependency on Feature::Compat::Try

perlpunk commented:
Please split this up in several PRs. There should be one PR for replacing Try::Tiny and a separate one for replacing eval.
I already found multiple problems and don't want do go over that big PR again and again.

okurz commented:
> Please split this up in several PRs. There should be one PR for replacing Try::Tiny and a separate one for replacing eval. I already found multiple problems and don't want do go over that big PR again and again.

No worries, done. I split it into three PRs now: #2650, #2651, #2652 and multiple different commits. As you like I can reshuffle this and split more. This PR was created with the intention to show how the complete picture would look like and if we like that.

Martchus reviewed (changes requested):
I wouldn't change all occurrences at once. It makes more sense to try it out in new/changed code first so we get some experience and see whether we actually like this.

I'm very much in favor of getting rid of `Try::Tiny`, though. This module is horrible as it makes the meaning of `return` statements completely unintuitive for everyone coming from other programing languages. Is `Feature::Compat::Try` better in that regard? If yes that's good but then your PR needs some further adjustments (see inline comment).

perlpunk commented:
> I'm very much in favor of getting rid of `Try::Tiny`, though. This module is horrible as it makes the meaning of `return` statements completely unintuitive for everyone coming from other programing languages (or plain `eval` usage). Is `Feature::Compat::Try` better in that regard?

Here is my comparison addressing things like that:
https://progress.opensuse.org/issues/176475

okurz commented on OpenQA/Commands.pm:
```diff
@@ -20,10 +20,10 @@ sub pass_message_from_ws_client_to_isotovideo ($self, $id, $msg) {
-        $app->log->warn('cmdsrv: failed to decode message');
+    catch ($e) {
+        $app->log->warn("cmdsrv: failed to decode message: $e");
         return undef;
```
No, this will return from the outer scope

okurz commented on autotest.pm:
```diff
@@ -266,20 +267,21 @@ sub _terminate () {    # uncoverable statement
-        $died = 1;    # test execution died
-    }
-    eval {
+    my $completed = do {
+        try { autotest::runalltests() }
+        catch ($e) {
+            warn $e;
+            $died = 1;    # test execution died
+        }
+    } // 0;
```
The `do` is what https://metacpan.org/pod/Feature::Compat::Try#try suggests if we want to use the result of the try scope. At first I also didn't like it but it's something I could get used to

okurz commented on OpenQA/Isotovideo/Utils.pm:
```diff
@@ -109,8 +109,10 @@ sub limit_git_cache_dir ($root_cache_dir, $current_cache_dir, $current_relative_
     my $index_lock = _lock_cache_directory($index_file);
-    my $index = eval { _open_cache_index($root_cache_dir, $index_file) };
-    die "Unable to open index for Git caching under '$index_file': $@" if $@;
+    my $index = do {
+        try { _open_cache_index($root_cache_dir, $index_file) }
+        catch ($e) { die "Unable to open index for Git caching under '$index_file': $e" }  # uncoverable statement
+    };
```
The do is wrong anyway here. I will change

perlpunk commented on bmwqemu.pm:
```diff
@@ -76,19 +77,19 @@ sub serialize_state (%state) {
+    try { open($fh, '<', $fn) }
+    catch ($e) { return 0 }
+    try { $ret = Cpanel::JSON::XS->new->relaxed->decode(<$fh>) }
+    catch ($e) { die "parse error in vars.json:\n$@" }    # uncoverable statement
```
See #2668 (`$e`)
